### DISCUSSION TITLE
Build a tree using a dataframe

### DISCUSSION THREAD

**Discussion by preinaj** [online-ml/river#1175] [Q&A] [answered]
Hi I'm currently trying to work with two tree models at the same time (two HoeffdingAdaptiveTreeRegressor to be precise), one would be training with the new instances that arrive from a streaming and the other one would be making predictions that are requested. After a certain time, I would like to update the predicting model with the parameters of the training model but I don't know if there is any way to do that. I've searched in the API references and found the to_dataframe method, but is there any way to use this data frame to build a new tree? Something like a to_tree method that recives a dataframe and returns a new tree.

Thanks in advance.

**Comment by MaxHalford** [comment, marked as answer]
Hey there!

The `to_dataframe` method is really just for introspection. At least, that was our intention. We haven't yet seen a lot of use cases for building a tree from a given structure, but I'll keep it in mind.

Could you not simply replace the model using the `clone` method? Something like this:

```py
predicting_model = training_model.clone()
```

**Reply by preinaj** [reply]
I thought that the clone method cloned the "base" model, with no seen instances. My idea was to duplicate the uptaded trained model, not the empty model that I created at the beginning of the training process. However, if the clone method allow me to do that, I think it could work fine.

**Reply by MaxHalford** [reply]
You're right, `clone` returns a fresh copy. My bad. However, you can use the `include_attributes` parameter to also copy over the model's state:

```py
predicting_model = training_model.clone(include_attributes=True)
```

Let me know if that works for you, as I haven't tested it with each model.

I will update the relevant documentation.

**Reply by preinaj** [reply]
```python
dataset = datasets.TrumpApproval()

model = tree.HoeffdingAdaptiveTreeRegressor(
         grace_period=50,
         model_selector_decay=0.3,
         seed=0
     )

metric = metrics.MAE()



evaluate.progressive_val_score(dataset, model, metric)

model2 = model.clone(include_attributes=True)
```

Using this code gives a maximum recursion depth exceeded error.

**Reply by MaxHalford** [reply]
Are you sure? What version of River are you using? It ran fine for me with version 0.15. I literally copy/pasted this into a Jupyter notebook:

```py
from river import datasets, tree, metrics, evaluate

dataset = datasets.TrumpApproval()

model = tree.HoeffdingAdaptiveTreeRegressor(
    grace_period=50,
    model_selector_decay=0.3,
    seed=0
)

metric = metrics.MAE()

evaluate.progressive_val_score(dataset, model, metric)

model2 = model.clone(include_attributes=True)
```

**Reply by preinaj** [reply]
I have the same version on Ubuntu 20.04.2 and gives me a really long error. This is the last part of the error that gives.

  File "/usr/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.8/copy.py", line 264, in _reconstruct
    y = func(*args)
RecursionError: maximum recursion depth exceeded

**Reply by MaxHalford** [reply]
Maybe it's because of the Python version, I'm using 3.10.

**Reply by preinaj** [reply]
Just found the answer, it's was the maximum recursion depth. It's 1000 by default and increasing it to 1500 solves the problem.

```python
import sys
sys.setrecursionlimit(1500)
```

**Reply by MaxHalford** [reply]
Ok interesting! I guess your tree was very deep 🌲 🌲 🌲
